### PR TITLE
feat(auth): add relay authentication; fix initial login flash

### PR DESCRIPTION
**Pull request by damassi**
This updates our relay layer to support authenticated queries and additionally fixes the logged in page flash.

### CODE REVIEW

File: src/system/relay/middleware.ts [hunk 1]
```diff
@@ -1,20 +1,36 @@
...
-    ttl: 60 * 1000,
-  }),
-]
+export const getRelayMiddleware = (
+  user?: UserSessionData | null
+): Middleware[] => {
+  const authenticatedHeaders: Headers = user
+    ? {
+        "X-USER-ID": user.id,
+        "X-ACCESS-TOKEN": user.accessToken,
```
Comment (jonallured) [inline src/system/relay/middleware.ts:19]: I always get confused about why we have both of these headers - does the token include the id? What would be mean to have a mis-match? 🤔

**Review by jonallured — APPROVED** [review]
LGTM!! 🎉

File: src/system/relay/middleware.ts (hunk 1 above)
Comment (damassi) [reply src/system/relay/middleware.ts:19]: It is MP weirdness. Some queries look explicitly for that id header, not sure why. I suspect its just old stuff

File: src/pages/_app.page.tsx [hunk 2]
```diff
@@ -28,4 +34,9 @@ function MyApp({ Component, pageProps }: AppProps) {
   )
 }
 
-export default MyApp
+MyApp.getInitialProps = async (appContext: AppContext) => {
+  const appProps = await App.getInitialProps(appContext)
+  const user = await getUserFromCookie(appContext.ctx.req as NextApiRequest)
+  appProps.pageProps.user = user
+  return { ...appProps }
+}
```
Comment (anandaroop) [inline src/pages/_app.page.tsx:42]: Oh, so being part of `_app.tsx` — does this now force every page now to be ssr? I.e. I'm wondering if we don't have to recommend [this kind of paperwork](https://github.com/artsy/forque/pull/24/files) any more?

File: src/system/artsy-next-auth/auth/cookies.ts [hunk 3]
```diff
@@ -16,8 +16,8 @@ export function setTokenCookie(res: NextApiResponse, token: string) {
   res.setHeader("Set-Cookie", cookie)
 }
 
-export function getTokenCookie(req: NextApiRequest) {
-  const cookies = parseCookies(req)
+export async function getTokenCookie(req: NextApiRequest) {
+  const cookies = await parseCookies(req)
```
Comment (anandaroop) [inline src/system/artsy-next-auth/auth/cookies.ts:20]: `parseCookies` isn't async tho, right? Why is this await necessary?

Comment (damassi) [reply src/system/artsy-next-auth/auth/cookies.ts:20]: To fake out the event loop and let it resolve correctly 😅

File: src/pages/_app.page.tsx (hunk 2 above)
Comment (damassi) [reply src/pages/_app.page.tsx:42]: > if we don't have to recommend this kind of paperwork

I think you're right! (I was just assuming that every page would naturally have a root relay base query, so would be using it by default per the pattern)

File: src/system/artsy-next-auth/auth/cookies.ts (hunk 3 above)
Comment (anandaroop) [reply src/system/artsy-next-auth/auth/cookies.ts:20]: Sooo... if I remove this async/await and everything still seems to work... what am I missing? Or how can I try to make it break?

(It's a small point, but just in the interest of keeping the hacks to the bare minimum necessary)

Comment (damassi) [reply src/system/artsy-next-auth/auth/cookies.ts:20]: If you removed it I _think_ what you would see is [this](https://artsy.slack.com/archives/C0376CFU527/p1648572501052449?thread_ts=1648567911.436469&cid=C0376CFU527) but only after the SSR pass. 

Page loads fine, click to A, then click back to home, triggers error.

**Pull request merged**